README.md
<!--
# This file is automatically generated by a
# `metapak` module. Do NOT change it in
# place, your changes would be overriden.
-->


# sencrop-developer-platform
> The Sencrop Developer Platform website



[//]: # (::contents:start)

This repository contains the code that powers the [Sencrop Developers Platform](https://developers.sencrop.com).

This project is based on [NextJS](https://github.com/zeit/next.js/),
 a universal website renderer.

That said, we currently only use it for its
 static exports.

# Build
```sh
npm run build
npm run export
```

# Development
To start working on the project, run:
```sh
npm run dev
```

[//]: # (::contents:end)

# Useful resources
- [Changelog](./CHANGELOG.md)

# License
[MIT](https://github.com/sencrop/sencrop-developer-platform/blob/master/LICENSE.md)


CHANGELOG.md
<a name="1.9.0"></a>
# [1.9.0](https://github.com/sencrop/sencrop-developer-platform/compare/v1.8.0...v1.9.0) (2019-02-20)


### Bug Fixes

* **API Guide:** Fix payloads with old/deprecated infos ([b2d39ed](https://github.com/sencrop/sencrop-developer-platform/commit/b2d39ed))
* **Partners API:** Fix links format ([871b012](https://github.com/sencrop/sencrop-developer-platform/commit/871b012))


### Features

* **Partners API:** Add a way to retrieve your own token as a partner ([d0579e6](https://github.com/sencrop/sencrop-developer-platform/commit/d0579e6))



<a name="1.8.0"></a>
# [1.8.0](https://github.com/sencrop/sencrop-developer-platform/compare/v1.7.1...v1.8.0) (2018-10-02)



<a name="1.7.1"></a>
## [1.7.1](https://github.com/sencrop/sencrop-developer-platform/compare/v1.7.0...v1.7.1) (2018-09-28)


### Bug Fixes

* **API Guide:** Rename forgotten old measure ([e665eb4](https://github.com/sencrop/sencrop-developer-platform/commit/e665eb4))



<a name="1.7.0"></a>
# [1.7.0](https://github.com/sencrop/sencrop-developer-platform/compare/v1.6.0...v1.7.0) (2018-09-28)


### Features

* **Anchor component:** Add a component to link to docs sections easyer ([507242d](https://github.com/sencrop/sencrop-developer-platform/commit/507242d))
* **API:** use prod api endpoint to display doc reference ([f650081](https://github.com/sencrop/sencrop-developer-platform/commit/f650081))
* **Swagger:** Swagger url based on env ([a2193c4](https://github.com/sencrop/sencrop-developer-platform/commit/a2193c4))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/sencrop/sencrop-developer-platform/compare/v1.5.0...v1.6.0) (2018-05-15)


### Bug Fixes

* **Assets:** Fix assets ([a80d99b](https://github.com/sencrop/sencrop-developer-platform/commit/a80d99b)), closes [#16](https://github.com/sencrop/sencrop-developer-platform/issues/16)
* **Layout:** Attempt to fix static assets ([efad916](https://github.com/sencrop/sencrop-developer-platform/commit/efad916)), closes [#16](https://github.com/sencrop/sencrop-developer-platform/issues/16)


### Features

* **API Guide:** Add bash syntaxic coloration ([41b1d8a](https://github.com/sencrop/sencrop-developer-platform/commit/41b1d8a))
* **API Guide:** Add infos on badly named WIND_* measures ([2d87344](https://github.com/sencrop/sencrop-developer-platform/commit/2d87344)), closes [#17](https://github.com/sencrop/sencrop-developer-platform/issues/17)
* **API Guide:** Improve time buckets description and usage ([dcb9745](https://github.com/sencrop/sencrop-developer-platform/commit/dcb9745)), closes [#14](https://github.com/sencrop/sencrop-developer-platform/issues/14)
* **Partner API:** Add some insight on a possible error on token creation ([010924b](https://github.com/sencrop/sencrop-developer-platform/commit/010924b)), closes [#15](https://github.com/sencrop/sencrop-developer-platform/issues/15)
* **Partners API:** Add information on the module delegation flow ([72cbb77](https://github.com/sencrop/sencrop-developer-platform/commit/72cbb77)), closes [#18](https://github.com/sencrop/sencrop-developer-platform/issues/18)



<a name="1.5.0"></a>
# [1.5.0](https://github.com/sencrop/sencrop-developer-platform/compare/v1.4.2...v1.5.0) (2018-04-27)


### Features

* **API Reference:** Update the Swagger ([c72bc43](https://github.com/sencrop/sencrop-developer-platform/commit/c72bc43))



<a name="1.4.2"></a>
## [1.4.2](https://github.com/sencrop/sencrop-developer-platform/compare/v1.4.1...v1.4.2) (2018-03-21)


### Bug Fixes

* **API Guide:** Fix the ping url link ([97d0b9c](https://github.com/sencrop/sencrop-developer-platform/commit/97d0b9c))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/sencrop/sencrop-developer-platform/compare/v1.4.0...v1.4.1) (2018-03-20)



<a name="1.4.0"></a>
# [1.4.0](https://github.com/sencrop/sencrop-developer-platform/compare/v1.3.0...v1.4.0) (2018-03-13)



<a name="1.3.0"></a>

pages/guide.js
import Link from 'next/link'
import Layout from '../components/Layout'
import Anchor from '../components/Anchor'

import SyntaxHighlighter, { registerLanguage } from "react-syntax-highlighter/light"
import js from 'react-syntax-highlighter/languages/hljs/javascript'
import bash from 'react-syntax-highlighter/languages/hljs/bash'
import docco from 'react-syntax-highlighter/styles/hljs/docco'

registerLanguage('javascript', js);
registerLanguage('bash', bash);

const PING_ENDPOINT = 'https://api.sencrop.com/v1/ping';
const CURL_PING_CODE = `curl ${PING_ENDPOINT}
# Answers: {"pong":"pong"}
`;
const TOKEN_RETRIEVAL_CODE = `console.log({
  token: 'Bearer ' +
    JSON.parse(localStorage.state).user.token,
  userid: JSON.parse(localStorage.state).user.id,
})
// Prints: { token: 'Bearer xxxxxxxx', userId: 1664 }
`;
const CURL_SENCROP_DEVICES_CODE = `curl 'https://api.sencrop.com/v1/users/1664/devices'\
  -H "Authorization: Bearer xxxxx"
`;
const SENCROP_DEVICES_PAYLOAD = `{
  "items": [
    33,
    114711
  ],
  "devices": {
    33: {
      "id": "33",
      "accessPeriods": [{
        "role": "collaborator",
        "startDate": "2017-03-15T23:00:00.000Z"
      }],
      "modelId": 7,
      "userId": 1,
      "organisationId": 1,
      "identification": "SC999999",
      "contents": {
        "name": "Rain sensor 1"
      }
    },
    "114711": {
      "id": 114711,
      "accessPeriods": [{
        "role": "collaborator"
      }],
      "modelId": 7,
      "userId": 1,
      "organisationId": 1,
      "identification": "SC666999",
      "contents": {
        "name": "Rain sensor 2"
      }
    },
  },
  "devicesStatuses": {
    "33": {
      "id": 33,
      "contents": {
        "latitude": 43.7799,
        "longitude": 1.31287
      }
    },
    "114711": {
      "id": "114711",
      "contents": {
        "latitude": 43.7799,
        "longitude": 1.3128
      }
    },
  },
  "models": {
    "7": {
      "id": 7,
      "contents": {
        "name": "Raincrop",
        "externalDiameter": 0.206,
        "conception": "France - Lille",
        "manufacturing": "Europe-France",
        "calibration": "Ok",
        "weight": 3.5
      }
    }
  }
}
`;
const CURL_SENCROP_RAW_DATA_CODE = `curl "https://api.sencrop.com/v1/users/1664/devices/33/data/raw?size=100&beforeDate=2017-10-10T00:00:00Z&measures=RELATIVE_HUMIDITY,TEMPERATURE"\
  -H "Authorization: Bearer xxxxx"
`;
const SENCROP_RAW_DATA_PAYLOAD = `[
  {
    "date": "2017-10-09T23:54:07.000Z",
    "type": "RELATIVE_HUMIDITY",
    "value": 65.2,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:54:07.000Z",
    "type": "TEMPERATURE",
    "value": 15.100000000000001,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:39:07.000Z",
    "type": "RELATIVE_HUMIDITY",
    "value": 66.7,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:39:07.000Z",
    "type": "TEMPERATURE",
    "value": 14.8,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:24:06.000Z",
    "type": "TEMPERATURE",
    "value": 14.9,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:24:06.000Z",
    "type": "RELATIVE_HUMIDITY",
    "value": 65.7,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:09:07.000Z",
    "type": "RELATIVE_HUMIDITY",
    "value": 67.4,
    "discarded": false
  },
  {
    "date": "2017-10-09T23:09:07.000Z",
    "type": "TEMPERATURE",
    "value": 14.5,
    "discarded": false
  },
  {
    "date": "2017-10-09T22:54:06.000Z",
    "type": "TEMPERATURE",
    "value": 15,
    "discarded": false
  },
  {
    "date": "2017-10-09T22:54:06.000Z",
    "type": "RELATIVE_HUMIDITY",
    "value": 64.2,
    "discarded": false
  }
]
`;
const CURL_SENCROP_DATA_CODE = `# Get hourly aggregated data for two days
curl "https://api.sencrop.com/v1/users/1664/devices/33/data/hourly?beforeDate=2017-10-07T07:34:32.000Z&days=7&measures=WIND_DIRECTION,WIND_SPEED"\
  -H "Authorization: Bearer xxxxx"
# Get daily aggregated data for two days
curl "https://api.sencrop.com/v1/users/1664/devices/33/data/daily?beforeDate=2017-10-07T07:34:32.000Z&days=90&measures=WIND_DIRECTION,WIND_SPEED"\
  -H "Authorization: Bearer xxxxx"
`;
const SENCROP_DATA_PAYLOAD = `{
  "item": 33,
  "devices": {
    "33": {
      "id": 33,
      "accessPeriods": [{
        "role": "owner",
        "endDate": "2038-01-19T03:14:07.000Z"
      }],
      "modelId": 8,
      "userId": 1664,
      "identification": "WC666999",
      "contents": {
        "name": "My Windcrop"
      }
    }
  },
  "measures": {
    "interval": "1h",
    "data": [{
      "key": 1507186800000,
      "WIND_SPEED": {
        "value": 15.5
      },
      "WIND_DIRECTION": {
        "value": 262
      },
      "docCount": 4
    },
    {
      "key": 1507190400000,
      "WIND_SPEED": {
        "value": 19.416666666666668
      },
      "WIND_DIRECTION": {
        "value": 281
      },
      "docCount": 12
    },
    { '...': '...' },
    {
      "key": 1507359600000,
      "WIND_SPEED": {
        "value": 15.5
      },
      "WIND_DIRECTION": {
        "value": 222
      },
      "docCount": 6
    }]
  }
}
`;
const CURL_SENCROP_STATISTICS_CODE = `# Get device statistics for two days
curl "https://api.sencrop.com/v1/users/1664/devices/33/statistics?startDate=2017-01-01T00:00:00.000Z&endDate=2017-02-01T00:00:00.000Z&measures=WIND_DIRECTION,WIND_SPEED&patched=false"\
  -H "Authorization: Bearer xxxxx"
`;
const SENCROP_STATISTICS_PAYLOAD = `{
  "item": 33,
  "devices": {
    "33": {
      "id": 33,
      "accessPeriods": [{
        "role": "owner",
        "endDate": "2038-01-19T03:14:07.000Z"
      }],
      "modelId": 8,
      "userId": 1664,
      "identification": "WC666999",
      "contents": {
        "name": "My Windcrop"
      }
    }
  },
  "measures": {
    "interval": "1d",
    "data": [{
      "key": 1507186800000,
      "WIND_SPEED": {
        "value": 15.5
      },
      "WIND_DIRECTION": {
        "value": 262
      },
      "docCount": 4
    },
    {
      "key": 1507190400000,
      "WIND_SPEED": {
        "value": 19.416666666666668
      },
      "WIND_DIRECTION": {
        "value": 281
      },
      "docCount": 12
    },
    { '...': '...' },
    {
      "key": 1507359600000,
      "WIND_SPEED": {
        "value": 15.5
      },
      "WIND_DIRECTION": {
        "value": 222
      },
      "docCount": 6
    }]
  }
}
`;
const CURL_SENCROP_GEOBASED_CODE = `
# Get geobased statistics
curl "https://api.sencrop.com/v1/users/1664/statistics?startDate=2017-01-01T00:00:00.000Z&endDate=2017-02-01T00:00:00.000Z&latitude=37.234894&longitude=-115.81082&measures=WIND_SPEED,RAIN_FALL"\
  -H "Authorization: Bearer xxxxx"
# Get geobased fixed scale data
curl "https://api.sencrop.com/v1/users/1664/data/hourly?beforeDate=2017-01-01T00:00:00.000Z&days=3&latitude=37.234894&longitude=-115.81082&measures=WIND_SPEED,RAIN_FALL"\
  -H "Authorization: Bearer xxxxx"
`;
const SENCROP_GEOBASED_PAYLOAD = `{
  "measures": {
    "interval": "1d",
    "data": [{
      "key": 1507186800000,
      "WIND_SPEED": {
        "precision": 80,
        "value": 15.5
      },
      "RAIN_FALL": {
        "precision": 99,
        "value": 3
      },
      "docCount": 4
    },
    {
      "key": 1507190400000,
      "WIND_SPEED": {
        "precision": 80,
        "value": 19.416666666666668
      },
      "RAIN_FALL": {
        "precision": 99,
        "value": 0.5
      },
      "docCount": 12
    },
    { '...': '...' },
    {
      "key": 1507359600000,
      "WIND_SPEED": {
        "precision": 80,
        "value": 15.5
      },
      "RAIN_FALL": {
        "precision": 99,
        "value": 0
      },
      "docCount": 6
    }]
  }
}`;

const SENCROP_TIME_BUCKET_CODE = `
import API from 'sencrop-js-api-client';
import moment from 'moment-timezone';

API.getUserDeviceStatistics({
  userId: 86,
  deviceId: 33,
  authorization: 'Bearer yolo-token',
  startDate: moment().tz('Europe/Paris').startOf('year').toISOString(),
  endDate: moment().tz('Europe/Paris').endOf('year').toISOString(),
  measures: ['TEMPERATURE'],
}, {
  // Here goes any Axios request configuration override
  // See: https://github.com/mzabriskie/axios#request-config
  timeout: 40000,
})
.then({ data } => {
  console.log(data);
});
`;

const SENCROP_TIME_BUCKET_PAYLOAD = `
{
  "item": 33,
  "devices": {
    "33": {
      "id": 33,
      "accessPeriods": [
        {
          "id": 10279,
          "deviceId": 33,
          "userId": 86,
          "type": "owner",
          "startDate": "2018-02-14T15:15:00.000Z"
        }
      ],
      "modelId": 7,
      "userId": 1217,
      "organisationId": 1061,
      "previousDevicesIds": [
        1308
      ],
      "identification": "RC001664",
      "serial": "ABBACA",
      "contents": {
        "name": ""
      }
    }
  },
  "measures": {
    "interval": "month",
    "data": [
      {
        "key": 1512082800000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1514761200000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1517439600000,
        "TEMPERATURE": {
          "value": 0.7823753330795595
        },
        "docCount": 7881
      },
      {
        "key": 1519858800000,
        "TEMPERATURE": {
          "value": 5.929323899371062
        },
        "docCount": 8904
      },
      {
        "key": 1522533600000,
        "TEMPERATURE": {
          "value": 12.416400462962963
        },
        "docCount": 8640
      },
      {
        "key": 1525125600000,
        "TEMPERATURE": {
          "value": 14.189615931721207
        },
        "docCount": 4218
      },
      {
        "key": 1527804000000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1530396000000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1533074400000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1535752800000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1538344800000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1541026800000,
        "TEMPERATURE": {},
        "docCount": 0
      },
      {
        "key": 1543618800000,
        "TEMPERATURE": {},
        "docCount": 0
      }
    ]
  }
}
`;

const Index = () => (
  <Layout title={ 'API Guide' } description={
`A simple guide to help you using our API.`.replace(/\n/, '')
  } guide>
    <h1>API Guide</h1>
    <p>
      The Sencrop API allows you to retrieve
      our users informations and create value
      on top of it.
    </p>
    <p>
      In this guide we discuss some design
      choices in order to help you grasp our
      API internals quicker.
    </p>
    <h2><Anchor text="First contact" /></h2>
    <p>
      You can simply hit the Sencrop API
      by simply opening our ping endpoint
      in your browser:<br/>
      <a href={PING_ENDPOINT}
      >{PING_ENDPOINT}</a>
    </p>
    <p>
      You probably noticed we are using the
      https protocol so that informations
      transiting between your systems and
      the Sencrop ones remain confidential.
    </p>
    <p>
      You can also use the command line to
      ping our API with the help of curl:
    </p>
    <SyntaxHighlighter language="bash" style={docco}>{
      CURL_PING_CODE
    }</SyntaxHighlighter>
    <h2><Anchor text="Get your token" /></h2>
    <p>
      Most of our API requires you to pass
      a token via the Bearer HTTP mecanism.
      To interact with our servers, you will
      need to choose a way to get one.
    </p>
    <h3><Anchor text="Via OAuth2" /></h3>
    <p>
      We plan to integrate OAuth authentication
      that will provide you a way to obtain
      a token from our users. Subscribe to our
      mailing list to be informed of the OAuth2
      support.
    </p>
    <h3><Anchor text="Via the Partners API" /></h3>
    <p>
      The <Link href="/partners"><a>Partners
      API</a></Link> allows you to ask our
      users an access to their data in a
      privilegied way. You first need to meet
      us to get that access.
    </p>
    <h3><Anchor text="Via the Sencrop app" /></h3>
    <p>
      Otherwise, for your own account, you
      can use our API with your own token,
      generated on our application, for
      testing purposes.
    </p>
    <p>
      To grab a token, just connect to
      the <a href="http://app.sencrop.com"
      >Sencrop Web Application</a> and log
      in.
    </p>
    <p>
      We will soon provide you an interface
      to manage your applications tokens but
      in the meanwhile, you can open the
      developer console and copy/paste this
      snippet:
    </p>
    <SyntaxHighlighter language="javascript" style={docco}>{
      TOKEN_RETRIEVAL_CODE
    }</SyntaxHighlighter>
    <p>
      Great! You can now access your Sencrop data
      with the help of your token. The user id
      will also help since almost every protected
      endpoints requires your to fill it.
    </p>
    <p>
      Starting here, you can directly jump to
      the <Link href="/reference"><a>API
      reference</a></Link> and insert your token
      right in the Authorization field of the
      protected endpoints.
    </p>
    <p>
      Or just continue your journey through
      our API guide!
    </p>
    <h2><Anchor text="Listing your devices" /></h2>
    <p>
      Before retrieving your data you may want
      to simply list your own Sencrop Devices:
    </p>
    <SyntaxHighlighter language="bash" style={docco}>{
      CURL_SENCROP_DEVICES_CODE
    }</SyntaxHighlighter>
    <p>The result will look like this:</p>
    <SyntaxHighlighter language="json" style={docco}>{
      SENCROP_DEVICES_PAYLOAD
    }</SyntaxHighlighter>
    <p>
      The <code>items</code> property tells
      you the collection of devices you can access
      to. While the <code>devices</code> hash
      allows your to pickup the devices details.
    </p>
    <p>
      You may ask why using that format. It allows
      our payload to avoid repating the same
      informations several times while not
      requiring you to use a specific JSON loader.
      It also allows us to significantly reduce
      our app memory footprint by easing hash
      merges accross our various states.
    </p>
    <h2><Anchor text="Reading raw device data" /></h2>
    <p>
      Your devices regularly send meteoroligical
      data to our servers. You can get their
      direct output by simply requesting the
      following endpoint:
    </p>
    <SyntaxHighlighter language="bash" style={docco}>{
      CURL_SENCROP_RAW_DATA_CODE
    }</SyntaxHighlighter>
    <p>The result will look like this:</p>
    <SyntaxHighlighter language="json" style={docco}>{
      SENCROP_RAW_DATA_PAYLOAD
    }</SyntaxHighlighter>
    <p>
      The <code>date</code>, <code>type</code>
      and <code>value</code> fields are self explanatory.
      The <code>discarded</code> field is a bit more
      special, he means that our algorithm detected
      that the measure was wrong. It can be due to
      many different issues (hardware failure,
      bad installation, network failures etc...).
    </p>
    <p>
      Note that the `size` query parameter is
      mandatory and limited to a few values.
      Check out the API reference for more
      information.
    </p>
    <h2><Anchor text="Reading device data" /></h2>
    <p>
      The raw data is cool but you may want a bit
      more insight on the data you retrieve.
    </p>
    <p>
      For this purpose we brought to you two kind
      of endpoints.
    </p>
    <SyntaxHighlighter language="bash" style={docco}>{
      CURL_SENCROP_DATA_CODE
    }</SyntaxHighlighter>
    <p>The result will look like this:</p>
    <SyntaxHighlighter language="json" style={docco}>{
      SENCROP_DATA_PAYLOAD
    }</SyntaxHighlighter>
    <p>
      Each measures are aggregated depending on
      its nature. The wind, temperature, relative
      humidity are simple averages while the rain
      is a sum instead. The wind direction is a
      vector sum.
    </p>
    <p>
      Note that the `days` query parameter is
      mandatory and limited to a few values.
      Check out the API reference for more
      information.
    </p>
    <h3>Adaptive scale data</h3>
    <p>
      Here, what matters for you is to get
      data for a given period. In this case,
      we automatically choose the right scale
      in order for you to get best insights
      with no performance hint.
    </p>
    <SyntaxHighlighter language="bash" style={docco}>{
      CURL_SENCROP_STATISTICS_CODE
    }</SyntaxHighlighter>
    <p>
      Note the patched parameter that allows your to retrieve
      only not patched data (ie only real device data). Setting
      it to true or not specifying it will not return
      extrapolated leaking data refills.
    </p>
    <p>The result will look like this:</p>
    <SyntaxHighlighter language="json" style={docco}>{
      SENCROP_STATISTICS_PAYLOAD
    }</SyntaxHighlighter>
    <h2><Anchor text="Geobased data" /></h2>
    <p>
      The geobased data provides your insights around
      a given position instead of a single device.
      Indeed, during its usage, a device may be moved
      or shutdown. Also, some devices can provide you
      rain measures while others will instead give
      you rain data.
    </p>
    <p>
      For that purpose, you may prefer get available data
      for a given place in your Sencrop network.
    </p>

    <SyntaxHighlighter language="bash" style={docco}>{
      CURL_SENCROP_GEOBASED_CODE
    }</SyntaxHighlighter>
    <p>The result will look like this:</p>
    <SyntaxHighlighter language="json" style={docco}>{
      SENCROP_GEOBASED_PAYLOAD
    }</SyntaxHighlighter>
    <p>
      You can notice a new property
      called <code>precision</code> that tells
      you how precise is the result. Indeed,
      the devices on which are based the data
      may not be at the exact position you
      provided but instead in a ten kilometers
      ring around it.
    </p>
    <p>
      This precision field goes from 0 to 100
      and is based on statistical studies we
      made upfront. That precision may vary
      depending on the kind of measure. It
      may also change since we will probably
      reshape our statistical algorithms
      for more sharp results.
    </p>
    <p>
      Note that the `days` query parameter is
      mandatory and limited to a few values.
      Check out the API reference for more
      information.
    </p>
    <h2><Anchor text="Time buckets" /></h2>
    <p>
      Various measures are aggregated into time
      buckets. The buckets are computed according
      to the user&apos;s timezone per default.
      Here are the various intervals you
      may encounter:
    </p>
    <ul>
      <li>
        <strong>15m</strong>:
        the bucket key will point the start of the
        buckets with the first one beeing the first
        15 minutes part of fours containing the given
        start date,
      </li>
      <li>
        <strong>30m</strong>:
        the bucket key will point the start of the
        buckets with the first one beeing the first
        30 minutes part of fours containing the given
        start date,
      </li>
      <li>
        <strong>hour</strong>:
        the bucket key will point the start of days
        hours inside the given time interval,
      </li>
      <li>
        <strong>week</strong>:
        the bucket key will point the start of weeks
        between the provided start and end dates
        (starting on sunday),
      </li>
      <li>
        <strong>month</strong>:
        the bucket key will point the start of months
        for each month between the provided start and
        end dates,
      </li>
      <li>
        <strong>year</strong>:
        the bucket key will point the start of years
        for each year between the provided start and
        end dates.
      </li>
    </ul>
    <p>
      You have to be aware of the fact that time buckets are aligned
      on days/months/years boundaries in the user's timezone. So,
      if you need to get the current year statistics for a french user,
      you need to set <code>startDate/endDate</code> values in the
      <code>Europe/Paris</code> timezone.
    </p>
    <p>
      For instance, with JavaScript and using the <a href="https://momentjs.com/">MomentJS
      library</a> you would end up with some code looking like
      the following:
    </p>
    <SyntaxHighlighter language="javascript" style={docco}>{
      SENCROP_TIME_BUCKET_CODE
    }</SyntaxHighlighter>
    <p>The result will look like this:</p>
    <SyntaxHighlighter language="json" style={docco}>{
      SENCROP_TIME_BUCKET_PAYLOAD
    }</SyntaxHighlighter>
    <p>
      Beware that not aligning on the user's timezone may lead
      to incomplete buckets at boundaries of the returned data.
      Also, if you retrieve a bucket for the current month, you
      only have a partial bucket for that month since it has not
      ended yet. In the above result you can see that the may
      month is not terminated yet so the bucket has a limited
      number of measures (see the <code>docCount</code>{' '}
      property) and the following buckets are empty since they
      represent future months.
    </p>
    <h2><Anchor text="Units" /></h2>
    <p>
      The API returns values in fixed units.
      We use the{' '}
      <a href="https://en.wikipedia.org/wiki/International_System_of_Units">
        international system of units
      </a>{' '}
      where applicable so that you can convert
      it on you side with ease. Here are the
      various units we use:
    </p>
    <ul>
      <li><strong>RAIN_FALL</strong>: pluviometry measured in{' '}
        <a href="https://en.wikipedia.org/wiki/Rain#Measurement">millimeters (mm)</a>,
      </li>
      <li><strong>TEMPERATURE</strong>: temperature measured in{' '}
        <a href="https://en.wikipedia.org/wiki/Celsius">degree Celsius (°C)</a>,
      </li>
      <li><strong>RELATIVE_HUMIDITY*</strong>: relative humidity measured in{' '}
        <a href="https://en.wikipedia.org/wiki/Relative_humidity#Definition">percentage (%)</a>,
      </li>
      <li><strong>WIND_SPEED/WIND_GUST</strong>: wind measured in{' '}
        <a href="https://en.wikipedia.org/wiki/Kilometres_per_hour">kilometers per hour (km·h−1)</a>,
      </li>
      <li><strong>WIND_DIRECTION</strong>: wind direction angle with the North in{' '}
        <a href="https://en.wikipedia.org/wiki/Degree_(angle)">angular degrees (°)</a>,
      </li>
      <li><strong>WET_TEMPERATURE</strong>: wet bulb temperature in{' '}
        <a href="https://en.wikipedia.org/wiki/Wet-bulb_temperature">degree Celsius (°C)</a>.
      </li>
    </ul>
    <p>
      Please note that the old counter intuitive name (<code>WIND_MAX</code>,
      <code>WIND_MEAN</code>, <code>RH_AIR_H1</code>, <code>TEMP_AIR_H1</code>,
      <code>RAIN_TIC</code>) are still supported but deprecated. We will probably
      remove those measures in a near future.
    </p>
    <h2><Anchor text="Limits" /></h2>
    <p>
      Since we are in an early alpha publication
      of this API, we do not yet provide feedback
      on rate limitations and how to prevent it.
      You will not reach the limit until you make
      a hundred calls per minutes which should be
      sufficient.
    </p>
    <p>
      Contact us if you think you need a more
      intensive access to the data in the
      meanwhile.
    </p>
    <h2><Anchor text="What&apos;s next ?" /></h2>
    <p>
      So you read it all? Impressive! You are
      now in the best conditions to use our
      API! You can try it interactively with
      our <Link href="/reference"><a>API
      reference</a></Link> or directly dive
      into code with our JavaScript SDK.
      Check out our <Link href="/tools"><a>open-source
      tools</a></Link>!
    </p>
  </Layout>
)

export default Index
